### PR TITLE
Fix path comparison and include additional guidance when secrets detected

### PR DESCRIPTION
**Pull request by sharon-wang**
- Addresses: https://github.com/posit-dev/positron/issues/3413#issuecomment-2195084061

### Fixes in this PR

- Fix path comparison issue on Windows
- Note in console output that `detect-secrets` should be installed if it isn't already
    image

- Add some guidance on what actions to take when secrets are detected
    image

### QA Notes

See https://github.com/posit-dev/positron/pull/3618 for more detail on testing and some example secrets to try.

#### Suggested tests
- test the hook on Windows
- test the hook without detect-secrets installed
- some new output should show when secrets are detected, which guide the user to remove the secrets / mark them as false positives / consult the secrets README

### CODE REVIEW

File: build/detect-secrets.js [hunk 1]
```diff
@@ -41,10 +41,11 @@ const execSync = (command, stdio = ['inherit', 'pipe', 'inherit']) => {
...
 	const gitCommand = 'git rev-parse --show-toplevel';
 	printDebug(gitCommand, true);
-	const gitRepoRoot = execSync(gitCommand);
+	const gitRepoRoot = execSync(gitCommand).trim();
 	printDebug(gitRepoRoot);
-	printDebug('current working directory: ' + process.cwd());
-	if (process.cwd().trim() !== gitRepoRoot.trim()) {
+	const currentDir = process.cwd().trim();
+	printDebug('current working directory: ' + currentDir);
+	if (path.resolve(currentDir) !== path.resolve(gitRepoRoot)) {
```
Comment (sharon-wang) [inline build/detect-secrets.js:48]: this fixes the Windows path comparison issue

File: build/detect-secrets.js [hunk 2]
```diff
@@ -107,9 +118,13 @@ const detectSecretsScan = (args, stdio) => {
...
 try {
-	detectSecrets('--version');
+	// pipe the stdio so we print our own error message if detect-secrets is not installed
+	detectSecrets('--version', stdio = 'pipe', throwIfError = true);
 } catch (error) {
+	printDebug(error);
+	console.error(); // print newline
 	console.error(`${'Error:'.red} detect-secrets is not installed. Install detect-secrets with ${'pip install detect-secrets'.magenta} or ${'brew install detect-secrets'.magenta}.`);
+	console.error(); // print newline
 	process.exit(ExitCodes.DETECT_SECRETS_WRAPPER_ERROR);
```
Comment (sharon-wang) [inline build/detect-secrets.js:128]: this, along with the changes above to include `throwIfError`, ensure we'll print the note to install `detect-secrets` if applicable when the precommit hook is run

File: build/detect-secrets.js [hunk 3]
```diff
@@ -164,23 +181,37 @@ const runDetectSecretsHook = () => {
...
-			console.error(error.stdout.toString().red);
+			console.error(); // print newline
+			// prints the secret detection output from detect-secrets-hook
+			if (error.stdout) {
+				console.error(error.stdout.toString().red);
+			}
+			// print guidance on how to manage the possible secrets
+			console.error('Uh oh! If you have secrets in your code, please remove them before committing.'.magenta);
+			console.error(`If you are certain that these are false positives, see ${'build/secrets/README.md'.underline} for instructions on how to mark them as such.`.magenta);
+			console.error(); // print newline
```
Comment (sharon-wang) [inline build/detect-secrets.js:198]: added guidance on what to do if secrets are detected

**Review by sharon-wang — COMMENTED** [review]
I have some file formatting fixes to apply as well, but I'll do that after this has been reviewed.

File: build/detect-secrets.js [hunk 4]
```diff
@@ -107,9 +118,13 @@ const detectSecretsScan = (args, stdio) => {
...
 // Ensure that detect-secrets is installed
 try {
-	detectSecrets('--version');
+	// pipe the stdio so we print our own error message if detect-secrets is not installed
+	detectSecrets('--version', stdio = 'pipe', throwIfError = true);
 } catch (error) {
+	printDebug(error);
+	console.error(); // print newline
 	console.error(`${'Error:'.red} detect-secrets is not installed. Install detect-secrets with ${'pip install detect-secrets'.magenta} or ${'brew install detect-secrets'.magenta}.`);
+	console.error(); // print newline
```
Comment (nstrayer) [inline build/detect-secrets.js:127]: Not terribly important but you can avoid these extra calls. 

```suggestion
	console.error(`\n${'Error:'.red} detect-secrets is not installed. Install detect-secrets with ${'pip install detect-secrets'.magenta} or ${'brew install detect-secrets'.magenta}.\n`);
```

File: build/detect-secrets.js [hunk 5]
```diff
@@ -130,8 +145,10 @@ const baselineFileExists = () => {
 // Ensure that the baseline file exists and exit if it does not
 const ensureBaselineFileExists = () => {
 	if (!baselineFileExists()) {
-		console.error(`${'Error:'.red} Baseline file ${baselineFile.underline} does not exist.
-Run ${'node build/detect-secrets.js init-baseline'.magenta} to create it.`);
+		console.error(); // print newline
+		console.error(`Error: Baseline file ${baselineFile.underline} does not exist.`.red);
+		console.error(`Run ${'node build/detect-secrets.js init-baseline'.magenta} to create it.`);
+		console.error(); // print newline
```
Comment (nstrayer) [inline build/detect-secrets.js:151]: Another example. 
```suggestion
		console.error(
			`\nError: Baseline file ${baselineFile.underline} does not exist.\n`.red +
			`Run ${'node build/detect-secrets.js init-baseline'.magenta} to create it.\n`
		);
```

File: build/detect-secrets.js [hunk 6]
```diff
@@ -17,10 +17,10 @@ const colors = require('colors');
 
 // Enum for exit codes
 const ExitCodes = {
-	SUCCESS: 0, 				        // success
+	SUCCESS: 0,                         // success
 	FOUND_SECRETS_OR_BASELINE_ISSUE: 1, // detect-secrets-hook found secrets or encountered a baseline file issue
 	DETECT_SECRETS_WRAPPER_ERROR: 2,    // an error occurred in this script
-	DETECT_SECRETS_ERROR: 3,            // an error occurred in detect-secrets
+	DETECT_SECRETS_ERROR: 999,          // an error occurred in detect-secrets (detect-secrets status code is appended)
```
Comment (nstrayer) [inline build/detect-secrets.js:23]: What's the reason for updating this to `999`? Is `999+n` a special status code for the hygiene checks?

**Review by nstrayer — APPROVED** [review]
Looks good to me. Just one question about the detect-secrets status code and a note about potentially reducing the number of calls to `console.error()` (I'm 50-50 though on if they actually make the code better.)

File: build/detect-secrets.js (hunk 6 above)
Comment (sharon-wang) [reply build/detect-secrets.js:23]: Previously, I was using `DETECT_SECRETS_ERROR: 3` as a prefix for the actual status code returned by `detect-secrets`, but I was using the format `<prefix>_<detect-secrets-code>` (e.g. `3_1`) which is not an integer and thus invalid to return.

So `999` is just to make it more visually clear in the format `<prefix><detect-secrets-code>`, e.g. `9991`

Comment (sharon-wang) [reply build/detect-secrets.js:23]: `detect-secrets` also uses the status codes 0 and 1 and least, and I wanted to differentiate between this wrapper script returning status code 1 and `detect-secrets` returning status code 1 (which would be returned as 9991 by the wrapper)

Comment (nstrayer) [reply build/detect-secrets.js:23]: Can exit codes be that high? I get that it's a string here but eventually, it becomes an 8-bit integer, right?

Comment (sharon-wang) [reply build/detect-secrets.js]: Haha yeah maybe they can't be that high :) It seems like the limit for a status code might be 255? 

I just set it back to `3` and removed the appending of the `detect-secrets` status code. The console error output and debug prints will have that extra information if needed.

Thanks for checking this!!

**Pull request merged**